pinot-integration-tests/src/test/java/org/apache/pinot/server/realtime/ControllerLeaderLocatorIntegrationTest.java

```diff
@@ -0,0 +1,169 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.server.realtime;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import org.apache.helix.HelixManager;
+import org.apache.pinot.common.utils.CommonConstants.Helix;
+import org.apache.pinot.common.utils.helix.LeadControllerUtils;
+import org.apache.pinot.controller.ControllerConf;
+import org.apache.pinot.controller.ControllerStarter;
+import org.apache.pinot.controller.helix.ControllerTest;
+import org.apache.pinot.pql.parsers.utils.Pair;
+import org.apache.pinot.util.TestUtils;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+
+public class ControllerLeaderLocatorIntegrationTest extends ControllerTest {
+  private static long TIMEOUT_IN_MS = 1000_000L;
+  private HashMap<Integer, String> _partitionToTableMap;
+
+  @BeforeClass
+  public void setUp() {
+    startZk();
+    startController();
+
+    FakeControllerLeaderLocator.create(_helixManager);
+
+    findTableNamesForAllPartitions();
+  }
+
+  @Test
+  public void testControllerLeaderLocator() {
+    Set<String> resultSet = new HashSet<>();
+    FakeControllerLeaderLocator controllerLeaderLocator = FakeControllerLeaderLocator.getInstance();
+
+    // Before enabling lead controller resource, helix leader should be used.
+    validateResultSet(controllerLeaderLocator, resultSet, 1, "Failed to get only one pair of controller");
+
+    // Enable lead controller resource
+    enableResourceConfigForLeadControllerResource(true);
+
+    // After resource config is enabled, use the lead controller in the resource.
+    validateResultSet(controllerLeaderLocator, resultSet, 1, "Failed to get only one pair of controller");
+
+    ControllerConf secondControllerConfig = getDefaultControllerConfiguration();
+    secondControllerConfig.setControllerPort(Integer.toString(DEFAULT_CONTROLLER_PORT + 1));
+    ControllerStarter secondControllerStarter = new ControllerStarter(secondControllerConfig);
+    secondControllerStarter.start();
+
+    TestUtils
+        .waitForCondition(aVoid -> secondControllerStarter.getHelixResourceManager().getHelixZkManager().isConnected(),
+            TIMEOUT_IN_MS, "Failed to start the second controller");
+
+    // After starting a second controller, there should be two leaders for all the partitions.
+    validateResultSet(controllerLeaderLocator, resultSet, 2, "Failed to get two pairs of controllers.");
+
+    // Disable lead controller resource
+    enableResourceConfigForLeadControllerResource(false);
+
+    // After disabling lead controller resource, only helix leader should be used.
+    validateResultSet(controllerLeaderLocator, resultSet, 1, "Failed to get only one pair of controller");
+
+    // Mock the behavior that 60 seconds have passed.
```

Comment:
nit: "Mock time so it is beyond minimum time to invalidate leader cache"

Reply:
Updated